### PR TITLE
Add State Transitions Attribute

### PR DESCRIPTION
**Pull request by RaniSputnik**
State in CommerceTools isn't particularly useful for us unless we can also provide transitions between states (to better enable CS to make changes in the Merchant Center).

This PR adds the transition attribute into state.

### CODE REVIEW

**Comment by RaniSputnik** [comment]
Hey folks - I'm a little confused by the coverage report above. It seems to be reporting massive test coverage drops for files I haven't even touched. Happy to add some more tests, especially if I've missed anything out. Let me know if you need me to make any changes.

**Comment by mvantellingen** [comment]
Thanks for the pull request, and don't worry about the code coverage comment since we still need to fine tune that bit.

I've briefly discussed this PR with @davidweterings and there are some details we need to think about and we need to update the Python SDK for it as well since we use that as a mock endpoint.

We will come back to this asap!

**Comment by davidweterings** [comment]
I looked at the docs (https://docs.commercetools.com/http-api-projects-states.html), I think a small change might be needed. The docs for transitions:

```
If transitions is an empty list, it means the current state is a final state and no further transitions are allowed.
    If transitions is not set, the validation is turned off. When performing a transitionState update action, any other state of the same type can be transitioned to.
```

This means there are three states 1. not filled in (no validation) 2. empty list (this is a final state) 3. a list of states that are allowed. I'm wondering if the code handles all 3 scenarios correctly. Thoughts @RaniSputnik ?

**Comment by RaniSputnik** [comment]
Hey thanks @davidweterings I must've missed that in the docs - I'll revisit this and see what I can do about it.

**Comment by RaniSputnik** [comment]
I've done some experimentation here and I can't find a nice seamless way for us to solve the problem. That could very well be down to lack of Terraform experience though, so thought I'd share my thoughts here in case any of you folks can think of a better solution.

I have the following `main.tf` file:
```
provider "commercetools" {}

resource "commercetools_state" "review_unreviewed" {
  key = "review-unreviewed"
  type = "ReviewState"
  name = {
      en = "Unreviewed"
  }
  description = {
    en = "Not reviewed yet"
  }
  initial = true
  roles = ["ReviewIncludedInStatistics"]
}

resource "commercetools_state" "review_published" {
  key = "review-published"
  type = "ReviewState"
  name = {
      en = "Published"
  }
  description = {
    en = "Reviewed and published to the public"
  }
  transitions = []
}
```

I thought, in an ideal world, the first state: `review_unreviewed` should result in no workflow defined ie. `null` transitions in CommerceTools and the second state `review_published` should result in an empty list of transitions because it's _explicitly_ empty.

I can see that this difference is reflected in state:
```
{
  "version": 4,
  "terraform_version": "0.12.5",
  "serial": 27,
  "lineage": "b6701ea0-8792-82ff-f868-4029bb32cd3a",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "commercetools_state",
      "name": "review_published",
      "provider": "provider.commercetools",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "description": {
              "en": "Reviewed and published to the public"
            },
            "id": "78ff572e-83cf-426f-a7d7-8e15f5b10b5f",
            "initial": false,
            "key": "review-published",
            "name": {
              "en": "Published"
            },
            "roles": null,
            "transitions": [],
            "type": "ReviewState",
            "version": 1
          },
          "private": "bnVsbA=="
        }
      ]
    },
    {
      "mode": "managed",
      "type": "commercetools_state",
      "name": "review_unreviewed",
      "provider": "provider.commercetools",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "description": {
              "en": "Not reviewed yet"
            },
            "id": "9f979c95-b61c-4bd7-bbfe-d8513b417e41",
            "initial": true,
            "key": "review-unreviewed",
            "name": {
              "en": "Unreviewed"
            },
            "roles": [
              "ReviewIncludedInStatistics"
            ],
            "transitions": null,
            "type": "ReviewState",
            "version": 1
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
```

But when using the `*schema.ResourceData.GetOkExists` method it returns `exists = false` for each of these two states. Perhaps I misunderstood how this method works? But I expected it to return false for `review_unreviewed` and true for `review_published`. Maybe this is even a bug in Terraform...

Regardless, I think my solution will be to add a `final` boolean property. Even though it doesn't exist in CommerceTools, it seems like a good companion property to `initial`. Setting `final=true` would indicate that the list of transitions should be treated as _explicitly_ empty rather than _implicitly_ empty (the null case where no transitions are defined). What do you think?

**Comment by davidweterings** [comment]
I think being as close to the Commercetools API is usually preferred. 

Some pointers: `GetOkExists` is only for checking boolean values, so that should not be used here. `GetOk` returns two values: the value and if the value exists, I would assume this is enough info to check if the value exists. If it does not exist it should be excluded from the state draft. Perhaps similar logic can be used for the update action?

**Comment by RaniSputnik** [comment]
Discovered this is blocked by a change required in the CommerceTools Golang SDK https://github.com/labd/commercetools-go-sdk/pull/45

**Comment by RaniSputnik** [comment]
I've spent some more time investigating our Terraform constraints. Unfortunately, it seems `GetOk` suffers from the same problem as `GetOkExists` - it can't distinguish between the case where the `transitions` property is omitted vs the `transitions` property being explicitly set to an empty array. Without being able to detect the difference and without a change in Terraform core, we will need a separate property in order to handle this scenario.

---

Full diagnosis.

The logging code (inside `resource_state.go#resourceStateCreate`)
```
getTransitions := d.Get("transitions")
getOkTransitions, getOk := d.GetOk("transitions")
getOkExistsTransitions, getOkExists := d.GetOkExists("transitions")
log.Printf("Get('transitions') = %+v, (Get('transitions') == nil) = %t", getTransitions, getTransitions == nil)
log.Printf("GetOk('transitions') = %+v, (GetOk('transitions') == nil) = %t, exists = %t", getOkTransitions, getOkTransitions == nil, getOk)
log.Printf("GetOkExists('transitions') = %+v, (GetOkExists('transitions') == nil) = %t, exists = %t", getOkExistsTransitions, getOkExistsTransitions == nil, getOkExists)
```

Produces the following with `TF_TRACE=1` enabled:
```
# Transitions explicitly set to empty array
2019-07-24T13:59:37.568+0100 [DEBUG] plugin.terraform-provider-commercetools: 2019/07/24 13:59:37 Get('transitions') = &{F:0x19b7ae0 m:map[] once:{m:{state:0 sema:0} done:0}}, (Get('transitions') == nil) = false
2019-07-24T13:59:37.568+0100 [DEBUG] plugin.terraform-provider-commercetools: 2019/07/24 13:59:37 GetOk('transitions') = &{F:0x19b7ae0 m:map[] once:{m:{state:0 sema:0} done:0}}, (GetOk('transitions') == nil) = false, exists = false
2019-07-24T13:59:37.568+0100 [DEBUG] plugin.terraform-provider-commercetools: 2019/07/24 13:59:37 GetOkExists('transitions') = &{F:0x19b7ae0 m:map[] once:{m:{state:0 sema:0} done:0}}, (GetOkExists('transitions') == nil) = false, exists = false

# Transitions omitted from resource
2019-07-24T13:59:37.569+0100 [DEBUG] plugin.terraform-provider-commercetools: 2019/07/24 13:59:37 Get('transitions') = &{F:0x19b7ae0 m:map[] once:{m:{state:0 sema:0} done:0}}, (Get('transitions') == nil) = false
2019-07-24T13:59:37.569+0100 [DEBUG] plugin.terraform-provider-commercetools: 2019/07/24 13:59:37 GetOk('transitions') = &{F:0x19b7ae0 m:map[] once:{m:{state:0 sema:0} done:0}}, (GetOk('transitions') == nil) = false, exists = false
2019-07-24T13:59:37.569+0100 [DEBUG] plugin.terraform-provider-commercetools: 2019/07/24 13:59:37 GetOkExists('transitions') = &{F:0x19b7ae0 m:map[] once:{m:{state:0 sema:0} done:0}}, (GetOkExists('transitions') == nil) = false, exists = false
```

The point in the above logs (as you can see) is that we get the same results and thus can not distinguish between these two scenarios.

**Comment by RaniSputnik** [comment]
I'd like to propose that we merge this in in it's current state and open a separate issue for catching the no transitions vs empty transitions case. I think it's going to be pretty difficult to overcome and this PR does enable the use of transitions, so it does add value (for my team at least).

Even if we were to overcome the limitations of Terraform in this instance, there's still the problem of adding support upstream for empty transition lists in labd/commercetools-go-sdk#45 anyway. I think the better course of action is to open an issue for this and follow up once the upstream issues have been resolved. 

What do you folks think?

**Comment by SergeyKovalchuk** [comment]
@RaniSputnik "I think the better course of action is to open an issue for this and follow up once the upstream issues have been resolved" We are using terraform provider too and it would be really appreciated to merge this for current iteration and as a next step do another issue.

**Comment by ilyasf** [comment]
Hello folks, I'am also interesting in merging this request, cause we are using TF provider to add states transitions and do it manually right now. In addition to add some issues to be resolved in short terms would be better than do nothing. Thanks.

**Comment by RaniSputnik** [comment]
Hey guys any update on this? This is going to start blocking us soon so I may have to just fork the repo and start building releases ourselves. I'm happy to add any other docs you think we need in order to telegraph the limitations with this approach.

**Comment by SergeyKovalchuk** [comment]
Hello @davidweterings. May we merge this for current iteration and as a next step do another issue? @RaniSputnik We are also waiting for this update (We would not like to create the same logic out of CT, would be better to use CT OOTB transition functionality).

**Comment by davidweterings** [comment]
@SergeyKovalchuk @RaniSputnik @ilyasf hey! I just got back from vacation, I will look at fixing the PR checks this week so we can merge it :)

File: commercetools/resource_state_test.go [hunk 1]
```diff
@@ -43,6 +44,33 @@ func TestAccState_createAndUpdateWithID(t *testing.T) {
...
 				),
 			},
+			{
+				Config: testAccTransitionConfig(t, transition),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"commercetools_state.acctest-t1", "transitions.#", "1",
+					),
+					resource.TestCheckResourceAttr(
+						"commercetools_state.acctest-t1", "transitions.0", transition,
```
Comment (davidweterings) [inline commercetools/resource_state_test.go]: I've been trying to fix the testing framework (which requires updating the Python sdk). I cannot get this check to work. I tried this against an actual Commercetools environment as well, which doesn't work either. Can you remove this check?

**Comment by davidweterings** [comment]
@SergeyKovalchuk @RaniSputnik @ilyasf I just released version 0.18.0 https://github.com/labd/terraform-provider-commercetools/releases

**Pull request merged**